### ISSUE TITLE
Thinning Issue on Networks with Concatenation in Forward Operation

### ISSUE THREAD

**Issue by soumendukrg** [NervanaSystems/distiller#424]
I was performing **filter pruning** and **network thinning** for a bunch of different networks. However, for networks which performs some form of _concatenation_, like Inception_V3, Squeezenet, Densenet, thinning always end up with an error. Can anyone provide me with some guidance? 

**Note**: Squeezenet1_1 does not give any thinning error when doing channel pruning, however test accuracy falls to **0%** after thinning, while its  **54%** before thinning. (Sparsity before thinning: 19%, after thinning: 22%)

I think [filter pruning issue](https://github.com/NervanaSystems/distiller/issues/86) talks about similar problem and it says using group dependency does not solve this problem. The user modified the thinning.py file for his own network. How can we generalize this? Is there already another issue related to this topic?

Any help with regard to this issue is appreciated.

**Comment by nzmora** [comment]
Hi @soumendukrg,

There are a couple of things to notice:
1. Thinning removes channels/filters and this cannot be done arbitrarily.  There are data dependencies between layers which are explained [here](https://nervanasystems.github.io/distiller/tutorial-struct_pruning.html).
2. Pruning channels/filters does not prune the bias coefficients and it does not prune the relevant gamma and beta parameters (x_hat = (x - x_avg) / x_std;  x_bn = gamma * h_hat + beta), in cases of convolutional layers followed by batch-normalization. On the other hand, thinning physically removes both the bias of channels/filters that are removed, and also the gamma and beta coefficients of a batch-normalization layer.  This means that after thinning you need to perform a bit of fine-tuning, so that the network adjusts to the removal of the bias, gamma and beta.

Take a look at these 2 items and see if they explain the issues that you see.  If you still require help, post here the pruning schedule and command-line you are using and I'll help.

Cheers,
Neta

**Comment by soumendukrg** [comment]
I was finally able to perform channel pruning on inception successfully after selecting proper layers to prune. I observed pruning middle layers in a block doesn't break the model. Thanks.

**Comment by nzmora** [comment]
Glad to hear.  Is there anything that we can do to make this process easier for others?
